motionplan/rrtStarConnect.go

```diff
@@ -83,138 +68,152 @@ func NewRRTStarConnectMotionPlannerWithSeed(
 	optimalCost := solutions[0].cost
 
 	// initialize maps
-	goalMap := make(map[node]node, len(solutions))
-	for _, solution := range solutions {
-		goalMap[newCostNode(solution.Q(), 0)] = nil
+	for i, solution := range solutions {
+		if i == 0 && mp.checkPath(rrt.planOpts, seed, solution.Q()) {
+			rrt.solutionChan <- &rrtPlanReturn{steps: []node{&basicNode{q: seed}, solution}}
+			mp.logger.Debug("RRT* could interpolate directly to goal")
+			return
+		}
+		rrt.rm.goalMap[newCostNode(solution.Q(), 0)] = nil
 	}
-	startMap := make(map[node]node)
-	startMap[newCostNode(seed, 0)] = nil
+	mp.logger.Debugf("RRT* failed to directly interpolate from %v to %v", seed, solutions[0].Q())
+	rrt.rm.startMap[newCostNode(seed, 0)] = nil
 
-	// for the first iteration, we try the 0.5 interpolation between seed and goal[0]
-	target := referenceframe.InterpolateInputs(seed, solutions[0].Q(), 0.5)
-
-	// Create a reference to the two maps so that we can alternate which one is grown
-	map1, map2 := startMap, goalMap
+	target := referenceframe.RandomFrameInputs(mp.frame, mp.randseed)
 
 	// Keep a list of the node pairs that have the same inputs
 	shared := make([]*nodePair, 0)
 
-	// sample until the max number of iterations is reached
-	var solutionCost float64
+	m1chan := make(chan node, 1)
+	m2chan := make(chan node, 1)
+	defer close(m1chan)
+	defer close(m2chan)
+
+	solved := false
+
 	for i := 0; i < algOpts.PlanIter; i++ {
 		select {
 		case <-ctx.Done():
-			solutionChan <- &planReturn{err: ctx.Err()}
+			// stop and return best path
+			if solved {
+				mp.logger.Debugf("RRT* timed out after %d iterations, returning best path", i)
+				rrt.solutionChan <- shortestPath(rrt.rm, shared, optimalCost)
+			} else {
+				mp.logger.Debugf("RRT* timed out after %d iterations, no path found", i)
+				rrt.solutionChan <- &rrtPlanReturn{planerr: ctx.Err(), rm: rrt.rm}
+			}
 			return
 		default:
 		}
 
 		// try to connect the target to map 1
-		if map1reached := mp.extend(algOpts, map1, target); map1reached != nil {
-			// try to connect the target to map 2
-			if map2reached := mp.extend(algOpts, map2, target); map2reached != nil {
-				// target was added to both map
-				shared = append(shared, &nodePair{map1reached, map2reached})
-			}
+		utils.PanicCapturingGo(func() {
+			mp.extend(algOpts, rrt.rm.startMap, target, m1chan)
+		})
+		utils.PanicCapturingGo(func() {
+			mp.extend(algOpts, rrt.rm.goalMap, target, m2chan)
+		})
+		map1reached := <-m1chan
+		map2reached := <-m2chan
+
+		if map1reached != nil && map2reached != nil {
+			// target was added to both map
+			solved = true
+			shared = append(shared, &nodePair{map1reached, map2reached})
 		}
 
 		// get next sample, switch map pointers
 		target = referenceframe.RandomFrameInputs(mp.frame, mp.randseed)
-		map1, map2 = map2, map1
-
-		// calculate the solution and log status of planner
-		if i%defaultSolutionCalculationPeriod == 0 {
```

Comment:
[minor] Do we not want this anymore?  This is the only way we have insight into the convergence.. but maybe thats not something we care about?